### PR TITLE
Home page (for connected users)

### LINKED ISSUE #120: User home page
- User information (success / defeats, blabla...)
- Challenges to play
- Pending challenges 

Home

### PR DESCRIPTION
**Pull request by valentinperignon**
Closes #120.

## Changes

- [x] User information (success / defeats, blabla...)
- [x] Challenges to play
- [x] Pending challenges
- [x] Updated [API documentation](https://github.com/ValFraNath/guacamole-api-docs)

Screen Shot 2021-01-27 at 00 28 45

### CODE REVIEW

File: client/src/pages/HomePage.js [hunk 1]
```diff
@@ -0,0 +1,169 @@
+import React, { Component } from "react";
+import axios from "axios";
+import { ChevronRightIcon } from "@modulz/radix-icons";
+
+import AuthService from "../services/auth.service";
+import Avatar from "../components/Avatar";
+import { Link } from "react-router-dom";
```
Comment (nathanaelhoun) [inline client/src/pages/HomePage.js:7]: Misplaced import

File: client/src/pages/HomePage.js [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+        <header>
+          <div id="header-background"></div>
+          <Avatar
+            size="125px"
+            eyes={currentUser?.avatar.eyes ?? 0}
+            hands={currentUser?.avatar.hands ?? 0}
+            hat={currentUser?.avatar.hat ?? 0}
+            mouth={currentUser?.avatar.mouth ?? 0}
+            colorBG={currentUser?.avatar.colorBG ?? "#d3d3d3"}
+            colorBody={currentUser?.avatar.colorBody ?? "#0c04fc"}
```
Comment (nathanaelhoun) [inline client/src/pages/HomePage.js:81]: If the props are required, I think they should not, and the `Avatar` should handle itself the default values

File: client/src/pages/HomePage.js [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+          </div>
+        </header>
+
+        <div id="links">
+          <Link to="/train" className="btn">
+            Entraînement
+          </Link>
+          <Link to="/homepage" className="btn">
+            Nouveau duel
+          </Link>
```
Comment (nathanaelhoun) [inline client/src/pages/HomePage.js:101]: We should visually mark that this doesn't work ATM

File: client/src/pages/HomePage.js [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+              {toPlayChallenges.map((value, index) => (
+                <article key={index}>
+                  <Avatar
+                    size="75px"
+                    eyes={usersData[value.opponent]?.avatar.eyes ?? 0}
+                    hands={usersData[value.opponent]?.avatar.hands ?? 0}
+                    hat={usersData[value.opponent]?.avatar.hat ?? 0}
+                    mouth={usersData[value.opponent]?.avatar.mouth ?? 0}
+                    colorBG={usersData[value.opponent]?.avatar.colorBG ?? "#d3d3d3"}
+                    colorBody={usersData[value.opponent]?.avatar.colorBody ?? "#0c04fc"}
```
Comment (nathanaelhoun) [inline client/src/pages/HomePage.js:121]: Same as above for the avatar

File: client/src/pages/HomePage.js [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+            Entraînement
+          </Link>
+          <Link to="/homepage" className="btn">
+            Nouveau duel
+          </Link>
+        </div>
+
+        <section>
+          <h2>
+            <img src={FightPilette} alt="Pilette is fighting" /> Ton tour
```
Comment (nathanaelhoun) [inline client/src/pages/HomePage.js:106]: Shouldn't alt texts be in French as all of our UI is in French?

File: client/src/pages/HomePage.js [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+                  </div>
+                </article>
+              ))}
+            </>
+          )}
+        </section>
+
+        <section>
+          <h2>
+            <img src={WaitPilette} alt="Pilette is waiting" /> En attente
```
Comment (nathanaelhoun) [inline client/src/pages/HomePage.js:138]:
```suggestion
            <img src={WaitPilette} alt="Pilette is waiting" /> En attente de l'adversaire
```

File: client/src/styles/pages/_HomePage.scss [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+
+      &:last-of-type {
+        margin-bottom: 0;
+      }
+
+      & > div {
+        &:first-of-type {
+          margin-right: 0.8rem;
+        }
+      }
```
Comment (nathanaelhoun) [inline client/src/styles/pages/_HomePage.scss:87]:
```suggestion
      & > div:first-of-type {
        margin-right: 0.8rem;
      }
```

File: server/controllers/user.js [hunk 8]
```diff
@@ -77,6 +77,54 @@ function queryCAS(login, pass) {
  * @apiSuccess {number} avatar.mouth      Number of the mouth
  */
 
+/**
+ * @api {post} /users/ Get data of several users
```
Comment (nathanaelhoun) [inline server/controllers/user.js:81]: Why is it a `post` endpoint?

File: server/controllers/user.js [hunk 9]
```diff
@@ -77,6 +77,54 @@ function queryCAS(login, pass) {
...
+        for (let value of sqlRes) {
+          usersData[value.pseudo] = {
+            pseudo: value.pseudo,
+            victories: Number(value.victories),
+            defeats: Number(value.defeats),
+            avatar: JSON.parse(value.avatar),
+          };
+        }
+      } catch (e) {
+        res.status(500).json({ error: e });
```
Comment (nathanaelhoun) [inline server/controllers/user.js:119]: For consistency
```suggestion
        res.status(500).json({ message: e });
```

File: server/controllers/user.js [hunk 10]
```diff
@@ -77,6 +77,54 @@ function queryCAS(login, pass) {
...
+      }
+      res.status(200).json(usersData);
+    })
+    .catch((error) => {
+      res.status(500).json({ error: error });
```
Comment (nathanaelhoun) [inline server/controllers/user.js:124]:
```suggestion
      res.status(500).json({ message: error });
```

File: client/src/pages/HomePage.js (hunk 3 above)
Comment (valentinperignon) [reply client/src/pages/HomePage.js:101]: Why? That's just for us, it will not be used by users

File: client/src/pages/HomePage.js (hunk 4 above)
Comment (valentinperignon) [reply client/src/pages/HomePage.js:121]: Yep :)

File: client/src/pages/HomePage.js (hunk 2 above)
Comment (valentinperignon) [reply client/src/pages/HomePage.js:81]: Yes!

File: server/controllers/user.js (hunk 8 above)
Comment (valentinperignon) [reply server/controllers/user.js:81]: The users whose information you wish to obtain are in an array passed by POST (as in the APIs given as an example).

File: client/src/pages/HomePage.js (hunk 5 above)
Comment (valentinperignon) [reply client/src/pages/HomePage.js:106]: Oh yes!

File: server/test/user.test.js [hunk 11]
```diff
@@ -91,6 +91,98 @@ describe("User test", function () {
     });
   });
 
+  describe("Get data from several users", function () {
+    let token;
+    before(function (done) {
+      // Authenticate
+      chai
+        .request(app)
```
Comment (FrancoisPog) [inline server/test/user.test.js:99]: You could use the requestAPI function of `index.test.js` but ok ! 👌

**Review by FrancoisPog — APPROVED** [review]
LGTM !

File: server/test/user.test.js (hunk 11 above)
Comment (valentinperignon) [reply server/test/user.test.js]: done()

File: server/controllers/user.js (hunk 8 above)
Comment (nathanaelhoun) [reply server/controllers/user.js:81]: It's true, the example API use this method. I'm wondering if we can use a `GET` endpoint with query parameters for the ids, as we are not likely to ask for a big number of users per request. I'm a bit bothered to use a `POST` endpoint to get data.

File: client/src/pages/HomePage.js (hunk 3 above)
Comment (nathanaelhoun) [reply client/src/pages/HomePage.js]: It's just for clarity. The alert box you add is perfect!

File: server/controllers/user.js (hunk 8 above)
Comment (valentinperignon) [reply server/controllers/user.js:81]: > as we are not likely to ask for a big number of users per request
I disagree, the user can have have as many challenges as he wants. We could request 2 users as 20, or more.

Comment (nathanaelhoun) [reply server/controllers/user.js:81]: Hum, you're right. Let's go for a `POST` endpoint :+1:

**Pull request merged**